### PR TITLE
refactor(webdriverio): rewrite webdriverio to TS

### PR DESCRIPTION
**Pull request by michael-siek**
See: #58, #75 

## Reviewer checks

**Required fields, to be filled out by PR reviewer(s)**

- [x] Code is reviewed for security

### CODE REVIEW

File: packages/webdriverio/.eslintrc.js [hunk 1]
```diff
@@ -1,15 +1,7 @@
 module.exports = {
   rules: {
-    '@typescript-eslint/no-empty-function': 'off',
+    '@typescript-eslint/no-explicit-any': 'off',
```
Comment (stephenmathieson) [inline packages/webdriverio/.eslintrc.js:3]: I don't think we should disable this rule. Using `any` defeats the purpose of TypeScript. Can we change this to a warning instead?

```suggestion
    '@typescript-eslint/no-explicit-any': 'warn',
```

**Review by stephenmathieson — COMMENTED** [review]
LGTM! Should get someone on your team to give the final approval tho 👍

File: packages/webdriverio/.eslintrc.js (hunk 1 above)
Comment (michael-siek) [reply packages/webdriverio/.eslintrc.js:3]: I could just remove this from eslint and add `// @ts-ignore` where I explicitly use `any` type which is only in test files.

Comment (stephenmathieson) [reply packages/webdriverio/.eslintrc.js:3]: If it's only in test files, then I'm strongly against turning this rule off globally. Just use an `override` instead:

```json
{
  "overrides": [
    {
      "files": ["*.test.ts", "*.test.tsx"],
      "rules": {
        "@typescript-eslint/no-explicit-any": "off"
      }
    }
  ]
}
```

Comment (stephenmathieson) [reply packages/webdriverio/.eslintrc.js:3]: I usually have something like this in my `eslintrc`:

```js
{
  "overrides": [
    {
      "files": ["*.test.ts", "*.test.tsx"],
      "rules": {
        // Tests can use `any`.
        "@typescript-eslint/no-explicit-any": "off",
        // Noop mocks, etc.
        "@typescript-eslint/no-empty-function": "off",
        // Tests can disable the TS compiler.
        "@typescript-eslint/ban-ts-comment": "off"
      }
    }
  ]
}
```

File: packages/webdriverio/README.md [hunk 2]
```diff
@@ -1,5 +1,140 @@
...
+  try {
+    const results = await builder.analyze();
+    console.log(results);
+  } catch (e) {
+    console.error(e);
+  }
+})();
+```
+
+## AxeBuilder({ client: WebdriverIO.BrowserObject })
```
Comment (jeeyyy) [inline packages/webdriverio/README.md:54]: Nitpick, all of these API methods should be nested under a level 2 heading API and the rest should be level 3 in my opinion.

File: packages/webdriverio/README.md [hunk 3]
```diff
@@ -1,5 +1,140 @@
...
+```js
+new AxeBuilder({ client }).include('.results-panel');
+```
+
+## AxeBuilder#exclude(selector: String)
+
+Add a CSS selector to the list of elements to exclude from analysis
+
+```js
+new AxeBuilder({ client }).include('.some-element').exclude('.another-element');
```
Comment (jeeyyy) [inline packages/webdriverio/README.md:75]:
```suggestion
new AxeBuilder({ client }).exclude('.another-element');
```

File: packages/webdriverio/README.md [hunk 4]
```diff
@@ -1,5 +1,140 @@
...
+```
+
+## AxeBuilder#options(options: [axe.RunOptions](https://github.com/dequelabs/axe-core/blob/develop/doc/API.md#options-parameter))
+
+Specifies options to be used by `axe.a11yCheck`. Will override any other configured options. including calls to `AxeBuilder#withRules()` and `AxeBuilder#withTags()`. See [axe-core API documentation](https://github.com/dequelabs/axe-core/blob/master/doc/API.md) for information on its structure.
```
Comment (jeeyyy) [inline packages/webdriverio/README.md:80]: Out of curiosity What is `axe.a11yCheck`?

File: packages/webdriverio/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+  private client: BrowserObject;
+  private axeSource: string;
+  private includes: string[];
+  private excludes: string[];
+  private option: RunOptions;
+  private disableFrameSelectors: string[];
+  constructor({ client }: Options) {
+    assert(
+      isWebdriverClient(client),
+      'An instantiated webdriverio client of v5 or v6 is required.'
```
Comment (jeeyyy) [inline packages/webdriverio/src/index.ts:28]: can we say this - https://github.com/dequelabs/axe-devtools-npm/blob/develop/packages/webdriverio/src/index.ts#L107

Given newer version 7, is coming out soon?

File: packages/webdriverio/src/utils/index.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  DoneFunction
+} from '../types';
+
+/**
+ * Validates that the client provided is WebdriverIO v5 or v6.
+ * @param {any} client
+ * @returns {BrowserObject}
+ */
+
+export const isWebdriverClient = (client: any): client is BrowserObject => {
```
Comment (jeeyyy) [inline packages/webdriverio/src/utils/index.ts:15]:
```suggestion
export const isWebdriverClient = (client: BrowserObject): boolean => {
```

File: packages/webdriverio/src/utils/index.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+ * @param {AnalyzePageParams} analyzeContext
+ * @param {DoneFunction} done
+ * @returns {AnalyzePageResponse}
+ */
+
+export const analyzePage = (
+  analyzeContext: AnalyzePageParams,
+  done: DoneFunction
+): AnalyzePageResponse => {
+  const axeCore = (window as any).axe;
```
Comment (jeeyyy) [inline packages/webdriverio/src/utils/index.ts:72]: You should do this - https://github.com/dequelabs/axe-devtools-npm/blob/develop/packages/cypress/src/types.ts#L15

File: packages/webdriverio/src/utils/index.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+/**
+ * Checks to make sure that the error thrown was not a stale iframe
+ * @param {Error} error
+ * @returns {void}
+ */
+
+export const logOrRethrowError = (error: Error): void => {
+  if (
+    (error as any).seleniumStack?.type === 'StaleElementReference' ||
```
Comment (jeeyyy) [inline packages/webdriverio/src/utils/index.ts:96]: nitpick, you could extend the Error interface same as global window instead of casting as any

**Review by jeeyyy — CHANGES_REQUESTED** [review]
Some feedback inline.

File: packages/webdriverio/README.md (hunk 4 above)
Comment (stephenmathieson) [reply packages/webdriverio/README.md]: It's the _old_ axe API. This was probably originally written years ago 🤷‍♂️

File: packages/webdriverio/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+  Element
+} from './types';
+import {
+  analyzePage,
+  isWebdriverClient,
+  normalizeContext,
+  logOrRethrowError
+} from './utils';
+
+export default class AxeBuilder {
```
Comment (stephenmathieson) [inline packages/webdriverio/src/index.ts:18]: Is it "safe" to change the export style here? The previous version used CommonJS-style exports I believe.

**Review by jeeyyy — APPROVED** [review]

**Review by jeeyyy — APPROVED** [review]

**Pull request merged**